### PR TITLE
feat: pnpm workspaces support

### PR DESCRIPTION
**Pull request by fahslaj**
## Description

Add support for pnpm workspaces:
- Detect workspace: protocol when building the package graph for commands.
- Properly update workspace: protocol dependencies during `lerna version`
- Prevent usage of `lerna bootstrap` and `lerna link` when using pnpm and direct the user to the pnpm docs. Instead, `pnpm install` should be used.
- Update the pnpm lockfile after `lerna version`

## Motivation and Context

https://github.com/lerna/lerna/issues/1818
https://github.com/lerna/lerna/issues/3195

This change allows users to use pnpm and lerna easily in the same project.

## How Has This Been Tested?

I tested changes manually with multiple test workspaces. Commands I tested manually include: version, publish, ls, run, bootstrap, link. I also added comprehensive unit tests around the new functionality and a couple end to end tests for versioning and publishing.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (change that has absolutely no effect on users)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: commands/add/index.js [hunk 1]
```diff
@@ -36,6 +36,13 @@ class AddCommand extends Command {
   }
 
   initialize() {
+    if (this.options.npmClient === "pnpm") {
+      throw new ValidationError(
+        "EPNPMNOTSUPPORTED",
+        "The 'add' command is not supported when using `pnpm` workspaces. Use `pnpm` directly to add dependencies to packages: https://pnpm.io/cli/add."
```
Comment (JamesHenry) [inline commands/add/index.js:42]: I think it's cleaner to not have the period after the URL (and maybe it would mess with terminals trying to make it clickable?)

```suggestion
        "The 'add' command is not supported when using `pnpm` workspaces. Use `pnpm` directly to add dependencies to packages: https://pnpm.io/cli/add"
```

File: commands/add/__tests__/add-command.test.js [hunk 2]
```diff
@@ -44,6 +44,15 @@ describe("AddCommand", () => {
     await expect(command).rejects.toThrow(/Requested package has no version:/);
   });
 
+  it("should throw when using pnpm", async () => {
+    const testDir = await initFixture("pnpm");
+    const command = lernaAdd(testDir)("@test/package-1");
+
+    await expect(command).rejects.toThrow(
+      "The 'add' command is not supported when using `pnpm` workspaces. Use `pnpm` directly to add dependencies to packages: https://pnpm.io/cli/add."
```
Comment (JamesHenry) [inline commands/add/__tests__/add-command.test.js:52]: I think it's cleaner to not have the period after the URL (and maybe it would mess with terminals trying to make it clickable?)

```suggestion
      "The 'add' command is not supported when using `pnpm` workspaces. Use `pnpm` directly to add dependencies to packages: https://pnpm.io/cli/add"
```

File: commands/bootstrap/__tests__/bootstrap-command.test.js [hunk 3]
```diff
@@ -138,6 +138,17 @@ describe("BootstrapCommand", () => {
     });
   });
 
+  describe("with pnpm", () => {
+    it("should throw validation error", async () => {
+      const testDir = await initFixture("pnpm");
+      const command = lernaBootstrap(testDir)();
+
+      await expect(command).rejects.toThrow(
+        "Bootstrapping with pnpm is not supported. Use pnpm directly to manage dependencies (https://pnpm.io/cli/install)."
```
Comment (JamesHenry) [inline commands/bootstrap/__tests__/bootstrap-command.test.js:147]: Please align to match formatting of error for `add`, e.g. when it comes to the URL

File: commands/bootstrap/index.js [hunk 4]
```diff
@@ -38,6 +38,13 @@ class BootstrapCommand extends Command {
   initialize() {
     const { registry, npmClient = "npm", npmClientArgs = [], mutex, hoist, nohoist } = this.options;
 
+    if (npmClient === "pnpm") {
+      throw new ValidationError(
+        "EWORKSPACES",
+        "Bootstrapping with pnpm is not supported. Use pnpm directly to manage dependencies (https://pnpm.io/cli/install)."
```
Comment (JamesHenry) [inline commands/bootstrap/index.js:44]: Please align to match formatting of error for `add`, e.g. when it comes to the URL

File: commands/link/__tests__/link-command.test.js [hunk 5]
```diff
@@ -192,44 +192,55 @@ Array [
...
+    });
+  });
+
+  describe("with pnpm", () => {
+    it("should throw validation error", async () => {
+      const testDir = await initFixture("pnpm");
+      const command = lernaLink(testDir)();
+
+      await expect(command).rejects.toThrow(
+        "Link is not supported with pnpm workspaces, since pnpm will automatically link dependencies during `pnpm install`. See the pnpm docs for details: https://pnpm.io/workspaces."
```
Comment (JamesHenry) [inline commands/link/__tests__/link-command.test.js:242]: Same comment regarding trailing period and overall consistency of error formatting for these

File: commands/link/index.js [hunk 6]
```diff
@@ -19,6 +20,13 @@ class LinkCommand extends Command {
   }
 
   initialize() {
+    if (this.options.npmClient === "pnpm") {
+      throw new ValidationError(
+        "EWORKSPACES",
+        "Link is not supported with pnpm workspaces, since pnpm will automatically link dependencies during `pnpm install`. See the pnpm docs for details: https://pnpm.io/workspaces."
```
Comment (JamesHenry) [inline commands/link/index.js:26]: Same comment regarding trailing period and overall consistency of error formatting for these

File: core/command/__tests__/command.test.js [hunk 7]
```diff
@@ -387,5 +387,24 @@ describe("core-command", () => {
...
+      await fs.writeJson(lernaConfigPath, {
+        ...lernaConfig,
+        useWorkspaces: false,
+      });
+
+      await expect(testFactory({ cwd })).rejects.toThrow(
+        expect.objectContaining({
+          prefix: "ENOWORKSPACES",
+          message:
+            "Usage of pnpm without workspaces is not supported. To use pnpm with lerna, set useWorkspaces to true in lerna.json and configure pnpm to use workspaces: https://pnpm.io/workspaces.",
```
Comment (JamesHenry) [inline core/command/__tests__/command.test.js:405]: Same comment regarding trailing period and overall consistency of error formatting for these

File: core/command/index.js [hunk 8]
```diff
@@ -247,6 +247,13 @@ class Command {
         `
       );
     }
+
+    if (this.options.npmClient === "pnpm" && !this.options.useWorkspaces) {
+      throw new ValidationError(
+        "ENOWORKSPACES",
+        "Usage of pnpm without workspaces is not supported. To use pnpm with lerna, set useWorkspaces to true in lerna.json and configure pnpm to use workspaces: https://pnpm.io/workspaces."
```
Comment (JamesHenry) [inline core/command/index.js:254]: Same comment regarding trailing period and overall consistency of error formatting for these

File: core/package-graph/__tests__/package-graph.test.js [hunk 9]
```diff
@@ -92,6 +92,240 @@ describe("PackageGraph", () => {
       expect(pkg1.localDependents.has("pkg-2")).toBe(true);
       expect(pkg2.localDependencies.has("pkg-1")).toBe(true);
     });
+
+    describe("with spec containing workspace: prefix", () => {
+      describe("localizes sibling when semver is satisfied", () => {
```
Comment (JamesHenry) [inline core/package-graph/__tests__/package-graph.test.js:97]: What do you mean by "localize" here sorry?

File: e2e/utils/fixture.ts [hunk 10]
```diff
@@ -130,10 +151,26 @@ export class Fixture {
...
+          `npx --registry=${REGISTRY} --yes lerna@${getPublishedVersion()} init ${args || ""}`
+        );
+        break;
+      case "pnpm":
+        execCommandResult = this.exec(`pnpm dlx lerna@${getPublishedVersion()} init ${args || ""}`);
+        break;
+      default:
+        throw new Error(`Unsupported package manager: ${this.packageManager}`);
+    }
+    return execCommandResult.then((initResult) =>
```
Comment (JamesHenry) [inline e2e/utils/fixture.ts:173]: Let's prefer async/await in new code please to aid with readability

File: e2e/utils/fixture.ts [hunk 11]
```diff
@@ -91,6 +93,25 @@ export class Fixture {
     return fixture;
   }
 
+  private async setNpmRegistry(packageManager: PackageManager): Promise<void> {
```
Comment (JamesHenry) [inline e2e/utils/fixture.ts:96]: The methods already have access to `this.packageManager`

File: e2e/utils/fixture.ts [hunk 12]
```diff
@@ -91,6 +93,25 @@ export class Fixture {
...
+    if (packageManager === "pnpm") {
+      await this.exec(`echo "registry=${REGISTRY}" > .npmrc`);
+    }
+  }
+
+  private async initializeNpmEnvironment(packageManager: PackageManager): Promise<void> {
```
Comment (JamesHenry) [inline e2e/utils/fixture.ts:102]: The methods already have access to `this.packageManager`

Comment (JamesHenry) [reply e2e/utils/fixture.ts:102]: Maybe `initializePackageManagerSettings` could be a more generic name for this method, and then the logic of the currently separate `setNpmRegistry` could be folded into it? Or do they need to be separate steps?

File: e2e/utils/fixture.ts [hunk 13]
```diff
@@ -22,7 +23,7 @@ interface FixtureCreateOptions {
   forceDeterministicTerminalOutput?: boolean;
 }
 
-type RunCommandResult = { stdout: string; stderr: string; combinedOutput: string };
+export type RunCommandResult = { stdout: string; stderr: string; combinedOutput: string };
```
Comment (JamesHenry) [inline e2e/utils/fixture.ts:26]: How come this was exported as part of the changes, I don't think it is used outside this file?

File: e2e/utils/fixture.ts [hunk 14]
```diff
@@ -1,17 +1,18 @@
 import { joinPathFragments, readJsonFile, writeJsonFile } from "@nrwl/devkit";
 import { exec, spawn } from "child_process";
-import { ensureDir, ensureDirSync, readFile, remove, writeFile } from "fs-extra";
+import { ensureDir, ensureDirSync, existsSync, readFile, remove, writeFile } from "fs-extra";
 import isCI from "is-ci";
+import { dump } from "js-yaml";
 import { dirSync } from "tmp";
 
-interface RunCommandOptions {
+export interface RunCommandOptions {
```
Comment (JamesHenry) [inline e2e/utils/fixture.ts:8]: How come this was exported as part of the changes, I don't think it is used outside this file?

**Review by JamesHenry — CHANGES_REQUESTED** [review]

**Comment by JamesHenry** [comment]
Great work @fahslaj! Excited about this

File: e2e/utils/fixture.ts (hunk 12 above)
Comment (fahslaj) [reply e2e/utils/fixture.ts:102]: I already tried this and had to split them out. setNpmRegistry has to happen first, since the rest of the initialization has to happen after running `lerna init`. The registry must be set first for the `npx` command that runs `lerna init`.

File: e2e/utils/fixture.ts (hunk 13 above)
Comment (fahslaj) [reply e2e/utils/fixture.ts:26]: RunCommandResult is returned by public functions of `Fixture`, so it should be exported too. I don't think it ever makes sense to allow the consuming code to obtain an instance of a type but have that type be private/not exported. 

This way, the consuming e2e tests can explicitly declare this type:
```
let commandOutput: RunCommandResult;
```

File: e2e/utils/fixture.ts (hunk 14 above)
Comment (fahslaj) [reply e2e/utils/fixture.ts:8]: `RunCommandOptions` and `RunCommandResult` are part of the public interface of `Fixture` because they are parameters / return values of Fixture's functions.

Comment (fahslaj) [reply e2e/utils/fixture.ts]: Removed the exports

File: e2e/utils/fixture.ts (hunk 13 above)
Comment (fahslaj) [reply e2e/utils/fixture.ts]: removed the exports

**Comment by lefreet** [comment]
look forward

**Review by JamesHenry — CHANGES_REQUESTED** [review]
Please see slack conversation for final testing plan before we land this

**Review by JamesHenry — APPROVED** [review]
Let's do this!

**Comment by AviVahl** [comment]
> Adds typescript as an explicit dependency of Lerna to satisfy strict peer dependency checking.

Could you clarify the reasoning here? how is lerna using typescript? is it using it via a third party?

**Comment by JamesHenry** [comment]
@AviVahl typescript is used at runtime for workspace package-graph analysis

**Comment by AviVahl** [comment]
Thanks. I've checked the code and this is only done by nx.

Installing nx, and now typescript, for everyone who uses lerna is beyond me. The direction this project has taken is quite worrisome from my perspective.
Others might not agree with the way I look at these things, but I'm personally starting to look for alternatives. :(

**Comment by JamesHenry** [comment]
Hi @AviVahl @petershershov @yardenporat 👋 

We can appreciate the instinctive concern, however please allow me to break down the full context here. Hopefully it will set your minds at ease.

Adding a dependency on Nx in order to use it within the implementation details of Lerna increases the size of the node_modules but only by a very small amount. An empty Lerna workspace without Nx is about 72MB, and an empty Lerna workspace with Nx is 75MB. It’s about a 5% increase for a completely empty workspace. For non-empty workspaces with real app and libs, the increase should be less than 1%. This is very small, and even this very small increase is a temporary one. By adding Nx, we will now be in a position to remove some older packages, so in the longer run the result will be smaller node_modules for all workspaces.

Lerna provides many capabilities: running tasks, bootstrapping, publishing, generating packages, importing existing packages. Unless you use all of those things, some percentage of the 72MB that Lerna had is already something you didn’t need or use. The small, temporary increase you see should therefore not be significant.

So that's the small, and temporary, trade-off covered, but what exactly does using Nx within some of lerna's implementation details enable for us and all lerna users?

Well, it's very, very significant:

1. One of the most-requested, longest-standing feature requests to Lerna is a dry-run mode for `lerna version` and `lerna publish`. Things like things are very hard to implement correctly. Thankfully Nx has such a capability baked in already, which has been battle-tested over many years. Thanks to Nx being there behind the scenes, we are going to be able to provide every single user with a `--dry-run` flag on `lerna version` which will show you what it will update (including `CHANGELOG.md` files), but without changing anything on disk until you are ready.
2. `lerna create` right now is not particularly useful or configurable, Nx's mature code generation allows us to rapidly improve on what `lerna create` can do (again with a reliable `--dry-run` capability "for free")
3. `lerna import`, a relatively complex command to reason about, can also be potentially be improved by showing what the result would be like without updating anything on disk.
4. It’s very hard to evolve lerna configuration right now because a lot of cleanup items would be breaking changes and it would be left to all users individually to deal with the consequences (just as with any OSS project). However, Nx has a built-in capability to autofix configurations, which we can leverage to get rid of legacy options in lerna.json and more. These improvements will be autofixed via a single command thanks to Nx being there in the implementation details.
5. Via Nx we can instantly provide a rich VSCode experience for Lerna users, including the built-in project graph visuazlier and other things like that. A lot of folks (e.g., RedwoodJS folks) have already been using this on their pure Lerna workspace and found it very useful for understanding how the workspace is structured and what will need to be refactored
6. Task running is faster and more ergonomic, with built-in distributable (vital for CI) caching

In summary then, leveraging Nx inside of lerna's implementation details should allow us to improve a lot of core capabilities of Lerna: code generation, importing, publishing, and running tasks - and also implement hotly requested, but otherwise complex to achieve, feature requests. All without requiring users to refactor their existing workspaces, or change their existing commands!

I hope that is all clear, if you have any questions about the above, or specific concerns about a particular package, or something is causing issues for you, please let us know and we will look into it for you.

Many thanks for being a part of the lerna community and for sharing your thoughts 🙏

**Comment by lefreet** [comment]
@fahslaj `lock` file not update after published. 

* so get some error in ci

```bash
 ERR_PNPM_OUTDATED_LOCKFILE  Cannot install with "frozen-lockfile" because pnpm-lock.yaml is not up to date with __packages__/ep-basic/package.json
Note that in CI environments this setting is true by default. If you still need to run install in such cases, use "pnpm install --no-frozen-lockfile"
```

* the same quastion is here  https://github.com/pnpm/pnpm/discussions/4241

**Comment by JamesHenry** [comment]
Hi @lefreet I commented on the discussion here: https://github.com/pnpm/pnpm/discussions/4241#discussioncomment-4281728

Please don't comment on old closed issues/PRs, it's not a scalable way to get support as it will likely be missed and/or be combined with stuff which isn't actually applicable to your specific concerns.

Please take a look at what I have written on the discussion and if you are still having trouble please open a new issue with a much information as possible about your specific situation.
